### PR TITLE
Convert WCF client instrumentation to a BindingElement

### LINKED ISSUE #478: WCF: Activity not stopped on EndpointNotFoundException (client-side)
# Issue with OpenTelemetry.Instrumentation.Wcf

**Environment (csproj):**

```csproj
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net472</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <Reference Include="System.ServiceModel" />
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="OpenTelemetry.Instrumentation.Wcf" Version="1.0.0-rc.6" />
  </ItemGroup>
</Project>
```

**Code:**

```cs
using System;
using System.Diagnostics;
using System.ServiceModel;

using OpenTelemetry;
using OpenTelemetry.Trace;
using OpenTelemetry.Instrumentation.Wcf;

[ServiceContract]
public interface ITestService
{
    [OperationContract]
    void Ping();
}

public static class Program
{
    public static void Main()
    {
        ActivitySource.AddActivityListener(new ActivityListener
        {
            ShouldListenTo = _ => true,
            Sample = (ref ActivityCreationOptions<ActivityContext> options) => ActivitySamplingResult.AllDataAndRecorded,
            ActivityStarted = _ => Console.WriteLine("Start"),
            ActivityStopped = _ => Console.WriteLine("Stop"),
        });

        using (Sdk.CreateTracerProviderBuilder().AddWcfInstrumentation().Build())
        {
            try
            {
                var factory = new ChannelFactory<ITestService>(
                    new BasicHttpBinding(),
                    new EndpointAddress("http://localhost/testservice"));
                factory.Endpoint.EndpointBehaviors.Add(new TelemetryEndpointBehavior());
                var channel = factory.CreateChannel();
                channel.Ping();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.GetType().Name);
            }

            Console.WriteLine($"Activity.Current is null? {Activity.Current is null}");
        }
    }
}
```

**Is this a feature request or a bug?**

* [ ] Feature Request
* [x] Bug

**What is the expected behavior?**

When running the code above without any service listening on "http://localhost/testservice", I expected the output to be:
```
Start
Stop
EndpointNotFoundException
Activity.Current is null? True
```

**What is the actual behavior?**

Actual output:
```
Start
EndpointNotFoundException
Activity.Current is null? False
```

## Additional Context

As you can see above, the `Activity` representing the outgoing request isn't stopped on `EndpointNotFoundException` (or any other exception thrown when there isn't any response from the server). Is this the expected behavior?

That is problematic because the stop event will never get logged/exported. Moreover, it might mess with surrounding code, which expected either to be no activity anymore or the previous activity to be restored.

I understand that OpenTelemetry.Instrumentation.Wcf uses `IClientMessageInspector.AfterReceiveReply()`, which is only called when _there is_ a reply. I wonder if there is any other extension point in WCF that we could use and that would be guaranteed to be called, even without a response.

### LINKED ISSUE #812: WCF client instrumentation on aspnet: Sequential WCF calls are incorrectly nested when async
# Issue with OpenTelemetry.Instrumentation.Wcf

Sequential async WCF call spans are nested inside the previous WCF call spans (their `Activity.ParentSpanId` is that of the activity for the previous WCF call) instead of the actual parent.

When running client instrumentation using OpenTelemetry.Instrumentation.Wcf, the `TelemetryClientMessageInspector` implementation of `AfterReceiveReply` is using the `correlationState` variable to work on the activity that was started in `BeforeSendRequest` and should encapsulate the WCF call. Unfortunately when the WCF call is async, `AfterReceiveReply` can run on an unrelated thread. The result is that when the activity has a parent (in my case due to the AspNet instrumentation but any activity parent will do), `activity.Stop()` will be operating on a different `Activity.Current` (in `System.Diagnostics`) when resetting to the previous Activity (the parent), this is because `Activity.Current` is an `AsyncLocal` static variable. This means that when you have a subsequent WCF call on the same thread as the first, `Activity.Current` will actually be the first WCF call and will be considered the parent of the new WCF call. 

I'm not really sure how this might be dealt with as its a combo of WCF behavior and `System.Diagnostics` use of `AsyncLocal` in the `Activity` class, maybe a check on whether the activity is stopped (that information is preserved)? 

The workaround is to wrap the WCF calls in another async task which ensures they share the static AsyncLocal Activity.Current variable as per this comment in a similar issue https://github.com/dotnet/extensions/issues/3412#issuecomment-676004005 but it's not lovely.

List of [all OpenTelemetry NuGet
packages](https://www.nuget.org/profiles/OpenTelemetry) and version that you are
using (e.g. `OpenTelemetry 1.0.2`):

OpenTelemetry.Instrumentation.Wcf v1.0.0.rc9.4 (but is persistent in latest version from source)
OpenTelemetry v1.3.1
OpenTelemetry.Api v1.3.1

Runtime version:

net472

**Is this a feature request or a bug?**

* [ ] Feature Request
* [x] Bug

**What is the expected behavior?**

That activity spans for sequential WCF calls wrapped in the same parent Activity will have the same `Activity.ParentSpanId` (Namely that of the parent Activity

**What is the actual behavior?**

The activity span for the second WCF call will reference the spanId of the first WCF call as `Activity.ParentSpanId` instead of the parent activity.

Here's a minimal example using a parent activity wrapping two consecutive WCF calls based on the aspnet WCF client example in the repo:

```csharp
using System;
using System.Diagnostics;
using System.ServiceModel;
using System.Threading.Tasks;
using OpenTelemetry;
using OpenTelemetry.Resources;
using OpenTelemetry.Trace;

namespace Examples.Wcf.Client;

internal static class Program
{
    private static readonly ActivitySource MyActivitySource = new ActivitySource(
        "MySource");

    public static async Task Main()
    {
        using var openTelemetry = Sdk.CreateTracerProviderBuilder()
            .SetResourceBuilder(ResourceBuilder.CreateDefault().AddService("Wcf-Client"))
            .AddSource("MySource")
            .AddAspNetInstrumentation()
            .AddWcfInstrumentation(x => x.SuppressDownstreamInstrumentation = false)
            .AddConsoleExporter()
            .Build();
        using (var activity = MyActivitySource.StartActivity("SayHello"))
        {
            activity?.SetTag("foo", 1);
            activity?.SetTag("bar", "Hello, World!");
            activity?.SetTag("baz", new int[] { 1, 2, 3 });
            activity?.SetStatus(ActivityStatusCode.Ok);
            await CallService("StatusService_Http").ConfigureAwait(false);
            //await CallService("StatusService_Tcp").ConfigureAwait(false);
        }

        Console.WriteLine("Press enter to exit.");
        Console.ReadLine();
    }


    private static async Task CallService(string name)
    {
        StatusServiceClient client = new StatusServiceClient(name);
        try
        {
            await client.OpenAsync().ConfigureAwait(false);

            var response = await client.PingAsync(
                new StatusRequest
                {
                    Status = Guid.NewGuid().ToString("N"),
                }).ConfigureAwait(false);
            var response2 = await client.PingAsync(
                new StatusRequest
                {
                    Status = Guid.NewGuid().ToString("N"),
                }).ConfigureAwait(false);
            Console.WriteLine($"Server returned: {response?.ServerTime}");
            Console.WriteLine($"Server returned next: {response2?.ServerTime}");
        }
        finally
        {
            try
            {
                if (client.State == CommunicationState.Faulted)
                {
                    client.Abort();
                }
                else
                {
                    await client.CloseAsync().ConfigureAwait(false);
                }
            }
            catch
            {
            }
        }
    }
}
```

### LINKED ISSUE #1153: Leaving 'SuppressDownstreamInstrumentation' as 'true' (default) breaks request telemetry
# Issue with OpenTelemetry.Instrumentation.Wcf

List of [all OpenTelemetry NuGet
packages](https://www.nuget.org/profiles/OpenTelemetry) and version that you are
using (e.g. `OpenTelemetry 1.3.2`):

OpenTelemetry.Exporter.OpenTelemetryProtocol  v1.4.0
OpenTelemetry.Exporter.OpenTelemetryProtocol.Logs  v1.4.0-rc.4
OpenTelemetry.Extensions.Hosting v1.4.0
OpenTelemetry.Instrumentation.AspNetCore v1.0.0-rc9.14
OpenTelemetry.Instrumentation.Runtime v1.0.0
OpenTelemetry.Instrumentation.Wcf v1.0.0-rc.9

Runtime version (e.g. `net462`, `net48`, `net6.0`, `net7.0` etc. You can
find this information from the `*.csproj` file):

NET6.0

**Is this a feature request or a bug?**

* [ ] Feature Request
* [x] Bug

**What is the expected behavior?**

I expected to see all my traces and logs associated with the request.

**What is the actual behavior?**

As soon as the first request is made using the instrumented WCF client, all other child scopes and logs are not propagated, and only the initiating scope is finished. 

Traces and logs without overriding `SuppressDownstreamInstrumentation` (true):

image

Traces and logs with explicit `.AddWcfInstrumentation(c => c.SuppressDownstreamInstrumentation = false)`:

image

I don't currently have a shareable repro project for this but the behavior is clearly linked to the instrumentation suppression logic. 

## Additional Context

I realized something was deeply wrong when I noticed even my logs were not being propagated after the first SOAP call. I could see that removing the WCF instrumentation would immediately change it and logs were working fine, then enabling WCF would only send whatever happened before the first SOAP call.

It then occurred to me to try disabling the suppression and it just worked... suddenly, everything was there again: all the missing logs, as well as all the other SOAP calls (we do several SOAP calls on the same request.

Additionally, all SOAP calls are `async`. The reason I mention that is because I saw there is an issue open that appears to be related to the fact that the calls are async:
- https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/812

### PR DESCRIPTION
**Pull request by repl-chris**
Fixes #478
Fixes #812
Fixes #1153

## Changes

This converts the WCF client instrumentation to being implemented in a lower-level bindingelement/channels rather than in the higher-level endpointbehavior/clientmessageinspector. This is needed so we can gain more insight into asynchronous execution flows and exception handling so we can better manage the ExecutionContexts. The clientmessageinspector still exists, but only to provide high-level context to the channels where the actual instrumentation now executes.

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by Kielek** [comment]
@repl-chris, please sing EasyCLA. Withot this we cannot proceed with this PR.

**Comment by repl-chris** [comment]
@Kielek yep, it is coming - it's working its way through corporate red tape as we speak 😀....I don't expect it will take long.

**Comment by repl-chris** [comment]
@Kielek ok good to go 👍

**Comment by repl-chris** [comment]
@CodeBlanch @Kielek Not meaning to rush, but is there any specific procedure for PRs? I just want to make sure I haven't missed a step or anything...

I realize it's a fairly large PR - I tried to keep the changes minimal but still functionally complete, but obviously it's a fairly fundamental change.  I could take a different approach if that would be preferable - I could do multiple smaller PRs adding the new components while keeping backwards compatibility before finally dropping the old mechanism?

**Comment by mconnew** [comment]
I haven't gone over the PR thoroughly, but wanted to give a bit of feedback. Using a BindingElement is going to be the most powerful way to achieve what's needed as you can get at layers that are otherwise really difficult to do. You can still keep the user experience based around adding a behavior without requiring them to modify their binding themselves. In the validate method of a behavior, you can get to the ServiceEndpoint and modify the binding. You would do this by creating a CustomBinding from the existing binding, add the custom BindingElement into the stack, then replace the Binding on the ServiceEndpoint with the new CustomBinding.  

There are some things to consider when using a binding element. It matters where you place it in the stack. If you put it just after the Transport and MessageEncoder, you will have access to the raw SOAP message, but there's 2 things to be aware of. If they are using SecurityMode.Message, it's possible you might not be able to read the headers as they might be encrypted. I believe the way you are adding the custom header that it probably won't be, but it's something to validate. The second thing is SOAP messages do map one to one with service calls. You will be exposed to all the infrastructure messages too such as reliable sessions keep alive pings, or session establishment handshaking. You may or may not want to have that logged. For example, there is some benefit to recording the rate of session establishment, or being aware of how often a reliable sessions is reestablished. This are both indicators of service health, but might be non-obvious on how to interpret that data, and it might need special handling. If you put it at the end of the stack, you will see messages intended to be dispatched. That will still capture some session establishment calls as internally it's implemented as another service implementation, but you won't see Reliable Sessions data at the end.  

Where you put the binding element will affect what the time taken to handle a request looks like. For example, if you are using security token authentication without using a session, every request needs to be authenticated. The authentication is done by the SecurityBindingElement. This authentication might involve an off-box call (eg SQL server lookup) and consume a not insignificant amount of time. If you place your binding element between the transport and the security binding elements, the request processing time you'd see would include the authentication time. If you placed your binding element at the end of the list, it won't capture that authentication timing. If something happens with authentication and it slows down and you are measuring from after authentication, you miss that in your metrics.  

Basically I'm saying you need to think carefully about where in the binding stack to place the binding element, or whether you want to insert multiple binding elements to capture more fine grained information. Do you want to emit data for non dispatching infrastructure messages?

**Comment by repl-chris** [comment]
@mconnew oooh thanks for the tip about replacing the binding on the serviceendpoint, I will definitely look into that. 

As for the placement of the binding element in the stack...that's some really good food for thought - I was thinking it should always go on the top (as far away from the transport as possible).  This telemetry is measuring the time to make an outgoing service call...I'd expect we'd want all the client-side WCF framework and extensions to be part of that, not simply just the network request (or a shade-of-grey between the 2 extremes).  Come to think of it, it would be nice to include all other endpointbehaviors inside of this too - logically they are also part of the "outgoing call"...I will play with this.  The transport already has its own instrumentation (at least when using http), and from a design perspective I'd expect if there are other elements of interest in the stack they should have instrumentation installed specifically for them....those could eventually also be part of this module, but for this initial implementation I would expect this component to just comprise as much of the WCF client-side stack as possible.

**Comment by repl-chris** [comment]
I have reworked it to inject the new bindingelement from the endpointbehavior as suggested, and it works beautifully.  It cleans it up a bunch - now there are no external api or config changes (yay!).

There's 1 specific area that I'm not very happy with....I wonder if you have any suggestions for a better approach.  In the `InstrumentedDuplexSessionChannel.cs` I start the activity on `Send()` and stop it on `Receive()`, but I have no good way to tie the 2 together - no way to pass any context from the Send call to the Receive call, so no easy way to know *which* activity to stop.  To make it work I'm manually tracking the outstanding requests (`RequestTelemetryStateTracker.cs`) so I can look up the appropriate request after receiving the response. Of course this only works reliably when I actually receive a response - any error on `Receive()` will cause the "outstanding requests" collection to accumulate entries (it'll hold max 1000 of them before it starts discarding them).  It's a little gross, but I still see this as a good step forward compared to where we are today, and of course it's just on the duplex channel - the request channel has no such hackery....I hope there's a better way 🤔

**Comment by mconnew** [comment]
The outgoing message has an ID in a special SOAP header. I'm on mobile right now so can't easily look it up but it should be easy to spot, it's a Guid. I think the header name is MessageID. The reply message will have another header I believe is named RelatesTo which contains the same Guid as the matching outgoing message. This is how WCF matches the up. We have a RequestReplyCorrelator which the outgoing message ID is added to before sending. Then we call a method on the correlator which returns (and when async, completes a pending async call) when a message is received with the matching RelatesTo header. We have a timeout mechanism which removes the outgoing ID from the correlator if the reply hasn't been seen without a timeout period. If a message is actually received after that timeout, the code drops the message as the matching ID is missing. I believe that timeout should be the Binding.SendTimeout. Any correlation you build should have a timeout mechanism and consider the call as failed if the reply isn't received within that timeout. The caller would have seen a TimeoutException. Anything received after that timeout should be treated as a failure. I don't know if you want to emit telemetry when that happens as you should have already emitted something saying it timed out, it would be useful info in a limited set of situations. That would generally happen if the server is configured with a longer timeout than the client.

**Comment by repl-chris** [comment]
I've revised it to use a timeout mechanism as suggested, and I'm much happier with it. Thanks 🙏

**Comment by mconnew** [comment]
Having a timer per entry is overly expensive. An alternative approach could be to include a deadline in the data that's stored, and then have a periodic timer which iterates through the list and removes any which have timed out. If you use an ordered set configured to iterate based on the deadline, iterating over the set to find expired items becomes really cheap.

**Comment by repl-chris** [comment]
Updated 👍

File: src/OpenTelemetry.Instrumentation.Wcf/Implementation/RequestTelemetryStateTracker.cs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+                }
+                else
+                {
+                    nextExpiry = ((Entry)kvp.Value).ExpiresAt;
+                    break;
+                }
+            }
+
+            timedOut?.ForEach(kvp => this.outstandingRequestStates.Remove(kvp.Key));
+            this.SetTimer(nextExpiry);
```
Comment (mconnew) [inline src/OpenTelemetry.Instrumentation.Wcf/Implementation/RequestTelemetryStateTracker.cs:123]: It's non-obvious that if there are no non-expiring items in outstandingRequestStates that nextExpiry will be null which will in effect cancel the timer. It might be worth adding a comment explaining this.

File: src/OpenTelemetry.Instrumentation.Wcf/Implementation/RequestTelemetryStateTracker.cs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        timedOut?.ForEach(kvp => this.TriggerTimedOutEvent((Entry)kvp.Value));
+    }
+
+    private void SetTimer(long? dueAt)
```
Comment (mconnew) [inline src/OpenTelemetry.Instrumentation.Wcf/Implementation/RequestTelemetryStateTracker.cs:129]: Instead of using `Nullable<long>` and having to check if it has a value or not and then retrieve that value, you could use `Timeout.Infinite`. When telling `System.Threading.Timer` that the timer is due in `Timeout.infinite`, it effectively turns off the timer. Then instead of disposing the existing timer and creating a new one, you could use `timer.Change(dueAt, Timeout.Infinite)` to modify the existing timer to fire at the new due time. You will need to wrap in a try/catch as `Timer.Change` throws `ObjectDisposedException` if there's a race and you dispose the Timer in the trackers Dispose method. Doing things this way would mean you would need to call `Timer.Dispose` in your Dispose method.

File: src/OpenTelemetry.Instrumentation.Wcf/Implementation/RequestTelemetryStateTracker.cs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    public void PushTelemetryState(Message request, RequestTelemetryState telemetryState)
+    {
+        var messageId = request?.Headers.MessageId?.ToString();
+        if (messageId != null)
+        {
+            var expiresAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() + this.configuredTimeoutInMs;
+            var entry = new Entry(telemetryState, expiresAt);
+            lock (this.sync)
+            {
+                this.outstandingRequestStates.Add(messageId, entry);
```
Comment (mconnew) [inline src/OpenTelemetry.Instrumentation.Wcf/Implementation/RequestTelemetryStateTracker.cs:50]: There is an obscure feature where you can cast an channel to an `IContextChannel` and set the `OperationTimeout` property to a different value, subsequent calls will use that timeout. You can't presume later requests will have a later deadline. You need to keep track of the current timers deadline and compare the new deadline to see if the timer needs to fire sooner.

Comment (repl-chris) [reply src/OpenTelemetry.Instrumentation.Wcf/Implementation/RequestTelemetryStateTracker.cs:50]: Ahhh shucks.  Y'know, I actually initially implemented it with the dynamic timeout functionality - rather than a timer/dict per channel I had a single global one for all channels (where each channel could have different timeouts configured)....but I couldn't find a good data structure to use.  Do you know of an available data structure which allows you to hash on 1 key (the message id) but sort by another (the deadline)?  In my initial implementation I ended up kinda implementing my own, and it came out a little gross so I figured a dict per-channel ordered by insertion order was cleaner....but if it needs to handle dynamic timeouts within a single channel then I guess that's not an option.

**Comment by repl-chris** [comment]
I've updated the "outstanding request tracker" to support dynamic timeout values, using a dict and a sortedset working in concert....I'm having a bit of trouble figuring out how to pull in the IContextChannel.OperationTimeout value though, any tips/tricks on how to read that value from within the channel?

**Comment by mconnew** [comment]
You don't want to be pulling the value from `IContextChannel.OperationTimeout` as that could have changed. E.g. if someone writes this code:
```c#
((IContextChannel)channel).OperationTimeout = TimeSpan.FromSeconds(20);
var requestTask = channel.DoSomethingAsync();
((IContextChannel)channel).OperationTimeout = TimeSpan.FromSeconds(10);
await requestTask;
```
Then the timeout value would likely be different by the time it gets to your code. The timeout value you care about is what's been passed to the channel Send methods such as `public void Send(Message message, , TimeSpan timeout)`. Getting the value from the OperationTimeout is done at a very early stage and is used by ServiceChannel when it calls Send with the actual Message object.

File: src/OpenTelemetry.Instrumentation.Wcf/Implementation/RequestTelemetryStateTracker.cs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+            this.TimeoutCallback = timeoutCallback;
+        }
+
+        public int CompareTo(object obj)
+        {
+            var other = (EntryTimeoutProperties)obj;
+            var result = this.ExpiresAt.CompareTo(other.ExpiresAt);
+            if (result == 0)
+            {
+                result = string.Compare(this.MessageId, other.MessageId, StringComparison.Ordinal);
```
Comment (mconnew) [inline src/OpenTelemetry.Instrumentation.Wcf/Implementation/RequestTelemetryStateTracker.cs:165]: This is unnecessary. A comparison result of 0 means they exist "in the same position" which in practive means the order they would appear in the SortedSet would be arbitrary. You don't care what order two items with the same expiry time will appear, so skip this extra calculation and return 0.

File: src/OpenTelemetry.Instrumentation.Wcf/Implementation/InstrumentedDuplexSessionChannel.cs [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    public IDuplexSession Session => this.Inner.Session;
+
+    private new IDuplexSessionChannel Inner { get => (IDuplexSessionChannel)base.Inner; }
+
+    public void Send(Message message)
+    {
+        this.SendInternal(message, _ => this.Inner.Send(message));
+    }
+
+    public void Send(Message message, TimeSpan timeout)
```
Comment (mconnew) [inline src/OpenTelemetry.Instrumentation.Wcf/Implementation/InstrumentedDuplexSessionChannel.cs:49]: This timeout is what you should be using for the telemetry tracking timeout. WCF channels created with a `ChannelFactory<TChannel>` shouldn't use the overload which has no timeout, but it's possible there's some edge cases where it doesn't (e.g. when used as an inner channel when using things like WS-Federation). To cover your basis, here's how you get the timeout when `Send(Message message)` is called. In `TelemetryBindingElement.BuildChannelFactory`, you need to capture the default send timeout from `BindingContext.Binding.SendTimeout` and pass that through to your instrumented channel factory implementations. From there, you need to pass the default send timeout to the Channel that you create. Use this as the timeout that you would pass to SendInternal.

Comment (mconnew) [reply src/OpenTelemetry.Instrumentation.Wcf/Implementation/InstrumentedDuplexSessionChannel.cs:49]: I just realized you already pass this value through. You would use this.telemetryTimeout when Send is called without a timeout, and you would use the timeout value pass in when Send is called with a timeout.

File: src/OpenTelemetry.Instrumentation.Wcf/Implementation/RequestTelemetryStateTracker.cs (hunk 4 above)
Comment (repl-chris) [reply src/OpenTelemetry.Instrumentation.Wcf/Implementation/RequestTelemetryStateTracker.cs:165]: `SortedSet<>` discards duplicates - if 2 requests happen to get the exact same expiry time then one of them would be lost and would never time out.  Strangely, `SortedList<>` also does not allow duplicates (it throws instead). The string comparison would be hit pretty rarely, I'd argue it's safer to leave it in.

File: src/OpenTelemetry.Instrumentation.Wcf/Implementation/InstrumentedDuplexSessionChannel.cs (hunk 5 above)
Comment (repl-chris) [reply src/OpenTelemetry.Instrumentation.Wcf/Implementation/InstrumentedDuplexSessionChannel.cs:49]: Hah sweet, thanks....seems obvious now that you point it out 😄

**Comment by repl-chris** [comment]
Updated with support for changed `OperationTimeout`

File: src/OpenTelemetry.Instrumentation.Wcf/TelemetryEndpointBehavior.cs [hunk 6]
```diff
@@ -62,8 +66,25 @@ public void Validate(ServiceEndpoint endpoint)
     {
     }
 
+    internal static void ApplyBindingElementToServiceEndpoint(ServiceEndpoint endpoint)
+    {
+        if (endpoint.Binding is CustomBinding customBinding && customBinding.Elements.Find<TelemetryBindingElement>() != null)
```
Comment (mconnew) [inline src/OpenTelemetry.Instrumentation.Wcf/TelemetryEndpointBehavior.cs:71]: If this runs on the service side too, you should probably also skip if `endpoint.IsSystemEndpoint` is `true`.

**Review by mconnew — APPROVED** [review]

File: src/OpenTelemetry.Instrumentation.Wcf/TelemetryEndpointBehavior.cs (hunk 6 above)
Comment (repl-chris) [reply src/OpenTelemetry.Instrumentation.Wcf/TelemetryEndpointBehavior.cs:78]: it won't currently ever run on the service-side, but it doesn't hurt to have it there 👍

**Review by lachmatt — APPROVED** [review]

**Comment by Kielek** [comment]
@repl-chris, could you please fix dotnet-format and markdownlint issues?

File: src/OpenTelemetry.Instrumentation.Wcf/README.md [hunk 7]
```diff
@@ -214,26 +210,7 @@ contracts you want to instrument:
...
-    catch (Exception)
-    {
-        var activity = Activity.Current;
-        if (activity != null && activity.Source.Name.Contains("OpenTelemetry.Instrumentation.Wcf"))
-        {
-            activity.Stop();
-        }
-    }
-```
+None
```
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.Wcf/README.md:213]: Remove whole Known issues section.

There is no common practice to keep it empty.

**Comment by repl-chris** [comment]
@Kielek all fixed up 👍

File: src/OpenTelemetry.Instrumentation.Wcf/Implementation/InstrumentedChannel.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.ServiceModel.Channels;
+
+namespace OpenTelemetry.Instrumentation.Wcf.Implementation;
+
+internal class InstrumentedChannel : InstrumentedCommunicationObject, IChannel
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.Wcf/Implementation/InstrumentedChannel.cs:21]: nit: Would you please add `sealed` to these `internal` classes. There is a slight perf boost for sealed classes. See: https://github.com/dotnet/runtime/issues/49944

File: src/OpenTelemetry.Instrumentation.Wcf/README.md [hunk 9]
```diff
@@ -50,8 +49,8 @@ using var openTelemetry = Sdk.CreateTracerProviderBuilder()
 
 ## WCF Client Configuration (.NET Framework)
 
-Add the `IClientMessageInspector` instrumentation via a behavior extension on
-the clients you want to instrument:
+Configure the `TelemetryEndpointBehaviorExtensionElement` on the clients
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.Wcf/README.md:52]: @repl-chris Do users need to do anything to their existing code to migrate to the new style when upgrading the package?

File: src/OpenTelemetry.Instrumentation.Wcf/Implementation/InstrumentedChannel.cs (hunk 8 above)
Comment (mconnew) [reply src/OpenTelemetry.Instrumentation.Wcf/Implementation/InstrumentedChannel.cs:21]: Sorry for the long comment, but I think this is a useful nugget of information to spread around.  

I don't think this would apply here as the call site of the virtual methods would be WCF code which has no visibility into the concrete type. The situation where you get the performance improvement is this:
```c#
public abstract class FooBase
{
    public abstract void DoWork()
}

public class MyFoo : FooBase
{
    public override void DoWork()
    {
        Console.WriteLine("Doing some work");
    }
}

public static class Program
{
    public static int Main(string[] args)
    {
        MyFoo foo = GetFoo();
        foo.DoWork();
    }
    public static MyFoo GetFoo()
    {
        return new MyFoo();
    }
}
```
In the snippet above, the code in `Main` method which calls `GetFoo()` doesn't have knowledge about if the `MyFoo` instance returned is an actual `MyFoo` instance as it could potentially be a type derived from `MyFoo`. As the `DoWork()` method is virtual, it could potentially have been overridden in a derived instance so it must do a virtual method call to make sure the correct method is called. If `MyFoo` was sealed, then you are guaranteed that the object returned from `GetFoo()` is an actual `MyFoo` type as you can't derive it. This means the call to `DoWork()` can be a non-virtual method call to `MyFoo.DoWork()` as there's no possibility that a different method implementation would get called. The non-virtual method call is faster. In this case it would probably inline the implementation as `MyFoo.DoWork()` is so small which gives you even more performance improvement.  

For completeness and accuracy, I ignored that the compiler might be able to still do it in this case as `GetFoo()` is non-virtual and short so could get inlined in which case the compiler might reason about the type sufficiently to do a non-virtual method call here. The point was about semantics and if `GetFoo()` was longer so couldn't be inlined, or virtual (it's static in this code, but generalizing here), then what I stated would be correct. Hopefully you understand the concept I'm explaining in the general case.  

In this codebase, the instrumented channel implements multiple WCF interfaces. The caller for all the virtual overridden methods (or interface implemented methods) is WCF which only has a reference to the instance as the interface type. There is no a priori knowledge when WCF was compiled about the type of the channel so the compiler can only emit virtual method calls.  

You only benefit from sealing an implementation if the call site has a reference typed as the concrete sealed type.  

On the other hand, marking it as sealed doesn't do any harm. Although not applicable in this scenario, sealing a class can also prevent using Ref.Emit to derive an implementation and using reflection to inject it to modify an app to do things not intended.

File: src/OpenTelemetry.Instrumentation.Wcf/README.md (hunk 9 above)
Comment (repl-chris) [reply src/OpenTelemetry.Instrumentation.Wcf/README.md:52]: No, there is no change needed to existing code to migrate to the new style - all the configuration stays the same.  I changed this README because it's not just a clientmessageinspector anymore, it is also a bindingelement.  But really, both of those details are irrelevant to someone using this library - as far as a user is concerned it's still just an endpointbehavior extension (just as it was before).

File: src/OpenTelemetry.Instrumentation.Wcf/Implementation/InstrumentedChannel.cs (hunk 8 above)
Comment (repl-chris) [reply src/OpenTelemetry.Instrumentation.Wcf/Implementation/InstrumentedChannel.cs:21]: I've added `sealed` to all possible internal classes 👍

**Review by Kielek — APPROVED** [review]

File: src/OpenTelemetry.Instrumentation.Wcf/README.md (hunk 9 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Instrumentation.Wcf/README.md:52]: Thanks for confirming!

**Review by CodeBlanch — APPROVED** [review]
LGTM thanks for the help on this @mconnew

**Comment by utpilla** [comment]
> Curious to know when we can expect a new version of the package with this fix in place. I'd like to use it right away!
> 
> @Kielek / @repl-chris , any idea?

@julealgon This change is available in `1.0.0-rc.11` version of the package: https://www.nuget.org/packages/OpenTelemetry.Instrumentation.Wcf/1.0.0-rc.11
